src/uu/pgrep/src/process.rs

```diff
@@ -0,0 +1,479 @@
+            _ => Err(io::ErrorKind::InvalidInput.into()),
+        }
+    }
+}
+
+impl TryFrom<&str> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        if value.len() != 1 {
+            return Err(io::ErrorKind::InvalidInput.into());
+        }
+
+        RunState::try_from(
+            value
+                .chars()
+                .nth(0)
+                .ok_or::<io::Error>(io::ErrorKind::InvalidInput.into())?,
+        )
+    }
+}
+
+impl TryFrom<String> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        RunState::try_from(value.as_str())
+    }
+}
+
+impl TryFrom<&String> for RunState {
+    type Error = io::Error;
+
+    fn try_from(value: &String) -> Result<Self, Self::Error> {
+        RunState::try_from(value.as_str())
+    }
+}
+
+/// Process ID and its information
+#[derive(Debug, Clone, Default)]
+pub struct ProcessInformation {
+    pub pid: usize,
+    pub cmdline: String,
+
+    inner_status: String,
+    inner_stat: String,
+
+    /// Processed `/proc/self/status` file
+    cached_status: Option<Rc<HashMap<String, String>>>,
+    /// Processed `/proc/self/stat` file
+    cached_stat: Option<Rc<Vec<String>>>,
+
+    cached_start_time: Option<u64>,
+    cached_tty: Option<Rc<HashSet<TerminalType>>>,
+}
+
+impl ProcessInformation {
+    /// Try new with pid path such as `/proc/self`
+    ///
+    /// # Error
+    ///
+    /// If the files in path cannot be parsed into [ProcessInformation],
+    /// it almost caused by wrong filesystem structure.
+    ///
+    /// - [The /proc Filesystem](https://docs.kernel.org/filesystems/proc.html#process-specific-subdirectories)
+    pub fn try_new(value: PathBuf) -> Result<Self, io::Error> {
+        let dir_append = |mut path: PathBuf, str: String| {
+            path.push(str);
+            path
+        };
+
+        let value = if value.is_symlink() {
+            fs::read_link(value)?
+        } else {
+            value
+        };
+
+        let pid = {
+            value
+                .iter()
+                .last()
+                .ok_or(io::ErrorKind::Other)?
+                .to_str()
+                .ok_or(io::ErrorKind::InvalidData)?
+                .parse::<usize>()
+                .map_err(|_| io::ErrorKind::InvalidData)?
+        };
+        let cmdline = fs::read_to_string(dir_append(value.clone(), "cmdline".into()))?
+            .replace('\0', " ")
+            .trim_end()
+            .into();
+
+        Ok(Self {
+            pid,
+            cmdline,
+            inner_status: fs::read_to_string(dir_append(value.clone(), "status".into()))?,
+            inner_stat: fs::read_to_string(dir_append(value, "stat".into()))?,
+            ..Default::default()
+        })
+    }
+
+    pub fn inner_status(&self) -> &str {
+        &self.inner_status
+    }
+
+    pub fn inner_stat(&self) -> &str {
+        &self.inner_stat
+    }
```

Comment:
I would rename those two functions because the "inner" in the function names is an implementation detail. The functions don't return an "inner status" or an "inner stat", they return the content of `/proc/<PID>/status` and `/proc/<PID>/stat`. And so my suggestion is to name them accordingly, something like `proc_status` and `proc_stat`. Or maybe you have better naming ideas ;-)

Reply:
Cool, I think `proc_*` would be better